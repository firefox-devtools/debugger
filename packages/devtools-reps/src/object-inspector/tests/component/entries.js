/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

 /* global jest */

const { mount } = require("enzyme");
const React = require("react");
const { createFactory } = React;
const ObjectInspector = createFactory(require("../../index"));
const { MODE } = require("../../../reps/constants");
const {
  formatObjectInspector,
  waitForDispatch,
  waitForLoadedProperties
} = require("../test-utils");
const gripMapRepStubs = require("../../../reps/stubs/grip-map");
const mapStubs = require("../../stubs/map");
const ObjectClient = require("../__mocks__/object-client");

function generateDefaults(overrides) {
  return {
    autoExpandDepth: 0,
    createObjectClient: grip => ObjectClient(grip),
    ...overrides
  };
}

function getEnumEntriesMock() {
  return jest.fn(() => ({
    iterator: {
      slice: () => mapStubs.get("11-entries")
    }
  }));
}

describe("ObjectInspector - entries", () => {
  it("renders Object with entries as expected", async () => {
    const stub = gripMapRepStubs.get("testSymbolKeyedMap");
    const enumEntries = getEnumEntriesMock();

    const oi = mount(ObjectInspector(generateDefaults({
      autoExpandDepth: 3,
      injectWaitService: true,
      roots: [{
        path: "root",
        contents: {value: stub}
      }],
      mode: MODE.LONG,
      createObjectClient: (grip) => {
        return ObjectClient(grip, {
          enumEntries,
        });
      },
      loadedProperties: new Map([
        ["root", mapStubs.get("properties")],
      ])
    })));

    const store = oi.instance().getStore();
    await waitForLoadedProperties(store, [
      "Symbol(root/<entries>/0)",
      "Symbol(root/<entries>/1)"
    ]);

    oi.update();
    expect(formatObjectInspector(oi)).toMatchSnapshot();

    // enumEntries shouldn't have been called since everything
    // is already in the preview property.
    expect(enumEntries.mock.calls.length).toBe(0);
  });

  it("calls ObjectClient.enumEntries when expected", async () => {
    const stub = gripMapRepStubs.get("testMoreThanMaxEntries");
    const enumEntries = getEnumEntriesMock();

    const oi = mount(ObjectInspector(generateDefaults({
      autoExpandDepth: 1,
      injectWaitService: true,
      roots: [{
        path: "root",
        contents: {
          value: stub
        }
      }],
      createObjectClient: (grip) => ObjectClient(grip, {enumEntries}),
      loadedProperties: new Map([
        ["root", {ownProperties: stub.preview.entries}],
      ]),
    })));

    expect(formatObjectInspector(oi)).toMatchSnapshot();

    const nodes = oi.find(".node");
    const entriesNode = nodes.at(1);
    expect(entriesNode.text()).toBe("<entries>");

    const store = oi.instance().getStore();
    let onEntrieLoad = waitForDispatch(store, "NODE_PROPERTIES_LOADED");
    entriesNode.simulate("click");
    await onEntrieLoad;
    oi.update();

    expect(formatObjectInspector(oi)).toMatchSnapshot();
    expect(enumEntries.mock.calls.length).toBe(1);

    entriesNode.simulate("click");
    expect(formatObjectInspector(oi)).toMatchSnapshot();

    entriesNode.simulate("click");

    expect(formatObjectInspector(oi)).toMatchSnapshot();
    // it does not call enumEntries if entries were already loaded.
    expect(enumEntries.mock.calls.length).toBe(1);
  });
});
