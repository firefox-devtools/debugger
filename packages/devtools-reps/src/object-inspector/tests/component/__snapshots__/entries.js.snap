// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectInspector - entries calls ObjectClient.enumEntries when expected 1`] = `
"
▼ Map(11)
|  ▶︎ <entries>
"
`;

exports[`ObjectInspector - entries calls ObjectClient.enumEntries when expected 2`] = `
"
  ▼ Map(11)
[ |  ▼ <entries> ]
  |  |  ▶︎ 0: \\"key-0\\" → \\"value-0\\"
  |  |  ▶︎ 1: \\"key-1\\" → \\"value-1\\"
  |  |  ▶︎ 2: \\"key-2\\" → \\"value-2\\"
  |  |  ▶︎ 3: \\"key-3\\" → \\"value-3\\"
  |  |  ▶︎ 4: \\"key-4\\" → \\"value-4\\"
  |  |  ▶︎ 5: \\"key-5\\" → \\"value-5\\"
  |  |  ▶︎ 6: \\"key-6\\" → \\"value-6\\"
  |  |  ▶︎ 7: \\"key-7\\" → \\"value-7\\"
  |  |  ▶︎ 8: \\"key-8\\" → \\"value-8\\"
  |  |  ▶︎ 9: \\"key-9\\" → \\"value-9\\"
  |  |  ▶︎ 10: \\"key-10\\" → \\"value-10\\"
"
`;

exports[`ObjectInspector - entries calls ObjectClient.enumEntries when expected 3`] = `
"
  ▼ Map(11)
[ |  ▶︎ <entries> ]
"
`;

exports[`ObjectInspector - entries calls ObjectClient.enumEntries when expected 4`] = `
"
  ▼ Map(11)
[ |  ▼ <entries> ]
  |  |  ▶︎ 0: \\"key-0\\" → \\"value-0\\"
  |  |  ▶︎ 1: \\"key-1\\" → \\"value-1\\"
  |  |  ▶︎ 2: \\"key-2\\" → \\"value-2\\"
  |  |  ▶︎ 3: \\"key-3\\" → \\"value-3\\"
  |  |  ▶︎ 4: \\"key-4\\" → \\"value-4\\"
  |  |  ▶︎ 5: \\"key-5\\" → \\"value-5\\"
  |  |  ▶︎ 6: \\"key-6\\" → \\"value-6\\"
  |  |  ▶︎ 7: \\"key-7\\" → \\"value-7\\"
  |  |  ▶︎ 8: \\"key-8\\" → \\"value-8\\"
  |  |  ▶︎ 9: \\"key-9\\" → \\"value-9\\"
  |  |  ▶︎ 10: \\"key-10\\" → \\"value-10\\"
"
`;

exports[`ObjectInspector - entries renders Object with entries as expected 1`] = `
"
▼ Map(2)
|    size: 2
|  ▼ <entries>
|  |  ▼ 0: Symbol(a) → \\"value-a\\"
|  |  |    <key>: Symbol(a)
|  |  |    <value>: \\"value-a\\"
|  |  ▼ 1: Symbol(b) → \\"value-b\\"
|  |  |    <key>: Symbol(b)
|  |  |    <value>: \\"value-b\\"
|  ▼ <prototype>: {…}
|  |    <prototype>: Object {  }
"
`;
